File: qt/src/core/Photographer.cc
#include "Photographer.hh"

#include <QDir>
#include <QRegularExpression>
#include <QtConcurrent>

#include "utils/Constants.hh"
#include "utils/PathUtils.hh"

namespace timelapse {

Photographer::Photographer(QObject* parent)
   : QObject(parent)
   , _screenCapture(ScreenCapture::create(this))
   , _imageProcessor(std::make_unique<ImageProcessor>(this))
   , _errorLog(std::make_unique<ErrorLog>(this)) {
   // Ensure directories exist
   PathUtils::ensureDir(PathUtils::timelapseDir());
   PathUtils::ensureDir(PathUtils::cacheDir());

   // Initialize database
   this->_database = std::make_unique<Database>(PathUtils::databasePath(), this);

   // Setup timer
   this->_captureTimer = new QTimer(this);
   this->_captureTimer->setSingleShot(true);
   connect(this->_captureTimer, &QTimer::timeout, this, &Photographer::captureScreenshot);
}

Photographer::~Photographer() {
   this->stop();
}

auto Photographer::isRunning() const -> bool {
   return this->_running.load();
}

auto Photographer::errorLog() const -> ErrorLog* {
   return this->_errorLog.get();
}

auto Photographer::database() const -> Database* {
   return this->_database.get();
}

auto Photographer::getScreenshotMetadata(uint32_t frameNumber)
   -> std::expected<std::optional<ScreenshotMetadata>, Error> {
   return this->_database->getScreenshotByFrame(frameNumber);
}

void Photographer::start() {
   if (this->_running.exchange(true)) {
      return;  // Already running
   }

   emit runningChanged(true);
   this->captureScreenshot();
}

void Photographer::stop() {
   if (!this->_running.exchange(false)) {
      return;  // Already stopped
   }

   this->_captureTimer->stop();
   emit runningChanged(false);
}

void Photographer::captureScreenshot() {
   if (!this->_running.load()) {
      return;
   }

   // Check if date changed (new day)
   QString today = PathUtils::currentDateString();
   if (today != this->_currentDate) {
      this->_currentDate = today;
      this->_currentDayDir.clear();  // Force recreation
   }

   // Ensure day directory exists
   auto dayDir = this->ensureDayDirectory();
   if (!dayDir) {
      this->logError(dayDir.error());
      this->scheduleNextCapture(kErrorWait);
      return;
   }

   // Capture screenshot
   auto image = this->_screenCapture->captureFocused();
   if (!image) {
      this->logError(Error{image.error().code(),
                           "Capture failed: " + image.error().message()});
      this->scheduleNextCapture(kErrorWait);
      return;
   }

   // Process image (resize with letterbox)
   auto processed = this->_imageProcessor->resizeWithLetterbox(
      *image, kTargetWidth, kTargetHeight);
   if (!processed) {
      this->logError(Error{processed.error().code(),
                           "Processing failed: " + processed.error().message()});
      this->scheduleNextCapture(kErrorWait);
      return;
   }

   // Check for black screen
   if (this->_imageProcessor->isImageBlack(*processed, kBlackThreshold)) {
      emit screenshotSkipped("Screen is black");
      this->scheduleNextCapture(kBlackScreenWait);
      return;
   }

   // Get next frame number
   auto frameNumber = this->getNextFrameNumber(*dayDir);
   if (!frameNumber) {
      this->logError(Error{frameNumber.error().code(),
                           "Failed to get frame number: " + frameNumber.error().message()});
      this->scheduleNextCapture(kErrorWait);
      return;
   }

   // Build filename
   QString filename = PathUtils::formatFrameNumber(*frameNumber) + ".png";
   QString filepath = *dayDir + "/" + filename;

   // Save image
   auto saveResult = this->_imageProcessor->saveImage(*processed, filepath);
   if (!saveResult) {
      this->logError(Error{saveResult.error().code(),
                           "Save failed: " + saveResult.error().message()});
      this->scheduleNextCapture(kErrorWait);
      return;
   }

   // Record in database
   QDateTime now = QDateTime::currentDateTimeUtc();
   QDateTime local = QDateTime::currentDateTime();
   auto dbResult = this->_database->insertScreenshot(*frameNumber, now, local);
   if (!dbResult) {
      this->logError(Error{dbResult.error().code(),
                           "Database insert failed: " + dbResult.error().message()});
      // Continue anyway - screenshot was saved
   }

   emit screenshotCaptured(filepath, *frameNumber);
   this->scheduleNextCapture(kScreenshotInterval);
}

auto Photographer::ensureDayDirectory() -> std::expected<QString, Error> {
   if (!this->_currentDayDir.isEmpty()) {
      return this->_currentDayDir;
   }

   QString dayDir = PathUtils::dayFolder(this->_currentDate);
   if (!PathUtils::ensureDir(dayDir)) {
      return std::unexpected(Error{ErrorCode::DirectoryCreationFailed,
                                   "Failed to create day directory: " + dayDir});
   }

   this->_currentDayDir = dayDir;
   return this->_currentDayDir;
}

auto Photographer::getNextFrameNumber(QString const& dayDir)
   -> std::expected<uint32_t, Error> {
   QDir dir(dayDir);
   if (!dir.exists()) {
      return 1;  // First frame
   }

   // Find highest existing frame number
   QStringList filters;
   filters << "*.png";
   QStringList files = dir.entryList(filters, QDir::Files, QDir::Name);

   if (files.isEmpty()) {
      return 1;  // First frame
   }

   // Parse the last filename to get the highest number
   static QRegularExpression re("^(\\d+)\\.png$");
   uint32_t maxFrame = 0;

   for (auto const& file : files) {
      auto match = re.match(file);
      if (match.hasMatch()) {
         uint32_t num = match.captured(1).toUInt();
         maxFrame = std::max(maxFrame, num);
      }
   }

   return maxFrame + 1;
}

void Photographer::scheduleNextCapture(std::chrono::milliseconds delay) {
   if (this->_running.load()) {
      this->_captureTimer->start(delay.count());
   }
}

void Photographer::logError(Error const& error) {
   this->_errorLog->append(error.message());
   emit errorOccurred(error.message());
}

}  // namespace timelapse

File: rust/src/core/photographer.rs
use chrono::{Local, Utc};
use std::fs;
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use crate::core::database::Database;
use crate::core::error::{Error, ErrorCode, Result};
use crate::core::image_processor::ImageProcessor;
use crate::core::screen_capture::ScreenCapture;
use crate::utils::constants::{
   BLACK_SCREEN_WAIT, BLACK_THRESHOLD, ERROR_WAIT, SCREENSHOT_INTERVAL, TARGET_HEIGHT, TARGET_WIDTH,
};
use crate::utils::paths;

pub struct Photographer {
   screen_capture: ScreenCapture,
   image_processor: ImageProcessor,
   database: Option<Database>,
   running: Arc<AtomicBool>,
   current_day_dir: Option<PathBuf>,
   current_date: Option<String>,
}

impl Photographer {
   pub fn new() -> Result<Self> {
      // Ensure directories exist
      paths::ensure_dir(&paths::timelapse_dir());
      paths::ensure_dir(&paths::cache_dir());

      // Initialize database
      let db_path = paths::database_path();
      let database = Database::new(&db_path).ok();

      Ok(Self {
         screen_capture: ScreenCapture::new(),
         image_processor: ImageProcessor::new(),
         database,
         running: Arc::new(AtomicBool::new(false)),
         current_day_dir: None,
         current_date: None,
      })
   }

   pub fn is_running(&self) -> bool {
      self.running.load(Ordering::SeqCst)
   }

   pub fn start(&mut self) {
      if self.running.swap(true, Ordering::SeqCst) {
         return; // Already running
      }

      let running = self.running.clone();

      // Run capture loop in current thread for simplicity
      // In a real app, you'd want this in a separate thread
      while running.load(Ordering::SeqCst) {
         match self.capture_screenshot() {
            Ok(Some(_)) => {
               thread::sleep(SCREENSHOT_INTERVAL);
            }
            Ok(None) => {
               // Black screen
               thread::sleep(BLACK_SCREEN_WAIT);
            }
            Err(e) => {
               eprintln!("Capture error: {}", e);
               thread::sleep(ERROR_WAIT);
            }
         }
      }
   }

   pub fn stop(&mut self) {
      self.running.store(false, Ordering::SeqCst);
   }

   /// Capture a single screenshot. Returns Ok(Some(path)) on success,
   /// Ok(None) if skipped (black screen), or Err on failure.
   pub fn capture_screenshot(&mut self) -> Result<Option<PathBuf>> {
      // Check if date changed
      let today = paths::current_date_string();
      if self.current_date.as_ref() != Some(&today) {
         self.current_date = Some(today);
         self.current_day_dir = None;
      }

      // Ensure day directory
      let day_dir = self.ensure_day_directory()?;

      // Capture screenshot
      let image = self.screen_capture.capture_focused()?;

      // Process image
      let processed = self
         .image_processor
         .resize_with_letterbox(&image, TARGET_WIDTH, TARGET_HEIGHT)?;

      // Check for black screen
      if self.image_processor.is_image_black(&processed, BLACK_THRESHOLD) {
         return Ok(None);
      }

      // Get next frame number
      let frame_number = self.get_next_frame_number(&day_dir)?;

      // Build filename
      let filename = format!("{}.png", paths::format_frame_number(frame_number));
      let filepath = day_dir.join(&filename);

      // Save image
      self.image_processor.save_image(&processed, &filepath)?;

      // Record in database
      if let Some(ref db) = self.database {
         let now_utc = Utc::now();
         let now_local = Local::now();
         if let Err(e) = db.insert_screenshot(frame_number, now_utc, now_local) {
            eprintln!("Database error: {}", e);
            // Continue anyway - screenshot was saved
         }
      }

      Ok(Some(filepath))
   }

   /// Get metadata for a specific frame
   pub fn get_screenshot_metadata(
      &self,
      frame_number: u32,
   ) -> Result<Option<crate::core::database::ScreenshotMetadata>> {
      match &self.database {
         Some(db) => db.get_screenshot_by_frame(frame_number),
         None => Ok(None),
      }
   }

   fn ensure_day_directory(&mut self) -> Result<PathBuf> {
      if let Some(ref dir) = self.current_day_dir {
         return Ok(dir.clone());
      }

      let date = self.current_date.as_ref().unwrap();
      let day_dir = paths::day_folder(date);

      if !paths::ensure_dir(&day_dir) {
         return Err(Error::new(
            ErrorCode::DirectoryCreationFailed,
            format!("Failed to create day directory: {:?}", day_dir),
         ));
      }

      self.current_day_dir = Some(day_dir.clone());
      Ok(day_dir)
   }

   fn get_next_frame_number(&self, day_dir: &PathBuf) -> Result<u32> {
      if !day_dir.exists() {
         return Ok(1);
      }

      let entries = fs::read_dir(day_dir).map_err(|e| {
         Error::new(
            ErrorCode::FileReadError,
            format!("Failed to read directory: {}", e),
         )
      })?;

      let mut max_frame: u32 = 0;

      for entry in entries.filter_map(|e| e.ok()) {
         let path = entry.path();
         if path.extension().map(|e| e == "png").unwrap_or(false) {
            if let Some(stem) = path.file_stem().and_then(|s| s.to_str()) {
               if let Ok(num) = stem.parse::<u32>() {
                  max_frame = max_frame.max(num);
               }
            }
         }
      }

      Ok(max_frame + 1)
   }
}

impl Default for Photographer {
   fn default() -> Self {
      Self::new().expect("Failed to create Photographer")
   }
}
